test/IssuanceFactory.js

```diff
@@ -40,239 +41,548 @@ const REFUND_ABI = {
+                now + 10, 1000000, THOUSAND_CLN, THOUSAND_CLN / 2,
+                'Some Name', 'SON', 18, CC_MAX_TOKENS, {from: owner})).logs[0].args.token;
+
+            cc = await ColuLocalCurrency.at(tokenAddress);
+        });
+
+        it('should create currency in a correct way', async () => {
+            let result = await factory.createIssuance(now + 10, 1000000, THOUSAND_CLN, THOUSAND_CLN / 2, 'Some Name', 'SON', 18, CC_MAX_TOKENS, {from: owner});
             assert.lengthOf(result.logs, 1);
             let event = result.logs[0];
             assert.equal(event.event, 'TokenCreated');
             tokenAddress = event.args.token
             assert(expect(tokenAddress).to.be.a('String'));
         });
 
-        it('should not be able to participate with 0 cln', async () => {
-            await expectRevert(Factory.participate(tokenAddress, 0, {from: accounts[1]}))
-        });
 
-        it('should not be able to participate with cln if no allowance was given', async () => {
-            await expectRevert(Factory.participate(tokenAddress, THOUSAND_CLN, {from: accounts[0]}))
+        describe('Participate in the CC issuance.', async () => {
+            it('should not be able to participate with 0 CLN', async () => {
+                await expectRevert(factory.participate(tokenAddress, 0, {from: notOwner}));
+            });
+
+            it('should not be able to participate before sale is open (for currency owner)', async () => {
+                cln.approve(factory.address, THOUSAND_CLN, {from: owner});
+                await expectRevert(factory.participate['address,uint256'](tokenAddress, THOUSAND_CLN, {from: owner}));
+            });
+
+            it('should not be able to participate before sale is open (for currency not owner)', async () => {
+                await cln.approve(factory.address, THOUSAND_CLN, {from: notOwner});
+                await expectRevert(factory.participate['address,uint256'](tokenAddress, THOUSAND_CLN, {from: notOwner}));
+            });
+
+            it('should not be able to participate with bad token address', async () => {
+                cln.approve(factory.address, THOUSAND_CLN, {from: notOwner})
+                await expectRevert(factory.participate['address,uint256']('0', THOUSAND_CLN, {from: owner}));
+                // using mmLib address instead of token address
+                await expectRevert(factory.participate['address,uint256'](mmLib.address, THOUSAND_CLN, {from: owner}));
+            });
+
+            it('should not be able to participate with CLN if no allowance was given', async () => {
+                // open the sale
+                await time.increaseTime(10);
+                await expectRevert(factory.participate(tokenAddress, THOUSAND_CLN, {from: owner}));
+            });
+
+            it('should be able to participate with CLN as owner if sale is open (approve, transfer)', async () => {
+                await time.increaseTime(10);
```

Comment:
you can add context for "sale not started", "Sale started", "Sale ended successfully" and "Sale failed" and do the same tests on all of the states.

Reply:
Done, made the test more organized